Client/Frontend/Settings/HomepageSettings/WallpaperSettings/v2/WallpaperSettingsViewModel.swift

```diff
@@ -0,0 +1,210 @@
+
+        let setWallpaperBlock = { [weak self] in
+            self?.updateCurrentWallpaper(for: wallpaper, in: collection) { result in
+                self?.selectedIndexPath = indexPath
+                completion(result)
+            }
+        }
+
+        if wallpaper.needsToFetchResources {
+            wallpaperManager.fetch(wallpaper) { result in
+                switch result {
+                case .success:
+                    setWallpaperBlock()
+                case .failure:
+                    completion(result)
+                }
+            }
+        } else {
+            setWallpaperBlock()
+        }
+    }
+}
+
+private extension WallpaperSettingsViewModel {
+
+    func setupWallpapers() {
+        wallpaperCollections = wallpaperManager.availableCollections
+    }
+
+    func cellViewModel(for wallpaper: Wallpaper,
+                       collectionType: WallpaperCollectionType,
+                       indexPath: IndexPath
+    ) -> WallpaperCellViewModel {
+        let a11yId = "\(a11yIds.card)_\(indexPath.section)_\(indexPath.row)"
+        var a11yLabel: String
+
+        switch collectionType {
+        case .classic:
+            a11yLabel = "\(stringIds.ClassicWallpaper) \(indexPath.row + 1)"
+        case .limitedEdition:
+            a11yLabel = "\(stringIds.LimitedEditionWallpaper) \(indexPath.row + 1)"
+        }
+
+        if wallpaperManager.currentWallpaper == wallpaper {
+            selectedIndexPath = indexPath
+        }
+
+        let cellViewModel = WallpaperCellViewModel(image: wallpaper.thumbnail,
+                                                   a11yId: a11yId,
+                                                   a11yLabel: a11yLabel)
+        return cellViewModel
+    }
+
+    func updateCurrentWallpaper(for wallpaper: Wallpaper,
+                                in collection: WallpaperCollection,
+                                completion: @escaping (Result<Void, Error>) -> Void) {
+        wallpaperManager.setCurrentWallpaper(to: wallpaper) { [weak self] result in
+            self?.setupWallpapers()
+
+            guard let extra = self?.telemetryMetadata(for: wallpaper, in: collection) else {
+                completion(result)
+                return
+            }
+            TelemetryWrapper.recordEvent(category: .action,
+                                         method: .tap,
+                                         object: .wallpaperSettings,
+                                         value: .wallpaperSelected,
+                                         extras: extra)
+
+           completion(result)
+        }
+    }
+
+    func telemetryMetadata(for wallpaper: Wallpaper, in collection: WallpaperCollection) -> [String: String] {
+        var metadata = [String: String]()
+
+        metadata[TelemetryWrapper.EventExtraKey.wallpaperName.rawValue] = wallpaper.id
+
+        let wallpaperTypeKey = TelemetryWrapper.EventExtraKey.wallpaperType.rawValue
+        switch wallpaper.type {
```

Comment:
do you think that will be easier to read as `switch (wallpaper.type, collection.type) ` to avoid having the nested switch and for default we could ignore collection.type value

Reply:
Smart!